### PR TITLE
Float interface improvements - and some other things

### PR DESCRIPTION
**Pull request by Joel-Dahne**
Lately I have been taking a look at interfacing Arblib with [IntervalArithmetic.jl](https://github.com/JuliaIntervals/IntervalArithmetic.jl). In this process I have implemented some functionality that would make this easier. Then I got stuck on a delayed train and ended up fixing less related things in the meantime...

This PR adds the following functionality:
- Adds `typemin` and `typemax` for `Mag`, `Arf` and `Arb`
- Adds `frexp` and `ldexp` for `Arf` and `Arb`. Arb already implements a `frexp` function, so this was easy!
- There is one large commit which improves the conversion to float types and `Complex{T}`, more comments about this below.
- Add support for `T == Float64` in `radius`. This doesn't allocate can be useful for example when heuristically choosing algorithm based on the radius. Since `Float64` has a bounded exponent this will overflow (rounding up) for extremely small or large radius.
- Add support for `T == Arf` in `getball`
- Microoptimization for `Irrational{:φ}`
- Add `log2` and `log10` for `Arb` and `Acb`. These are sometimes used for magnitude estimation in generic Julia code.
- Improve conversion to `Int` and `BigInt`. The conversion to `Int` would previously crash on overflow. One could possibly add conversion to other integer types, but they would just be wrappers around these types anyway. I also added a note about it being nice to support rounding methods such as `round`, `ceil` and `floor`.

Regarding the conversion to floats and complex: 

For `Float64` it now supports specifying the rounding as the second argument when converting. Before it would fall back to a generic Julia implementation if one specified the rounding and I'm not entirely sure that it was completely rigorous. The new version should be faster anyway. Previously we supported giving a `rnd::Arblib.arb_rnd` keyword argument to specify the argument. Since we are now at `1.0` I have opted to keep this implementation as well, though it could be removed at some future point. I also added `Float16` and `Float32` conversion, wrapping the `Float64` one. In particular Makie.jl likes to convert things to `Float32`.

For `BigFloat` it not supports giving the precision and the rounding as arguments in the same format that is mentioned in the docstring of `BigFloat`. Here there is however one design decision that I would like input on. The docstring for `BigFloat` specifies that the default precision is given by `precision(BigFloat)` and the rounding mode by `rounding(BigFloat)`. I have opted to have the precision be given by the precision of the input, and the rounding mode to always be `RoundNearest`. I very much prefer to avoid using the global arguments when possible, they tend to lead to a lot of issues. But contradicting the `BigFloat` documentation could of course also lead to confusion for people. Do you have any thoughts about this?

I have also thought a bit about how to handle conversion of `Arb` to floats. Currently it just converts the midpoint, which is in general what makes the most sense. If one specifies the rounding then this is applied when converting the midpoint. One could imagine the rounding the specify if one wants the midpoint, lower bound or upper bound of the ball. So that one would have something like `Float64(Arb((0, 1))) = 0.5`, `Float64(Arb((0, 1)), RoundDown) = 0.0` and `Float64(Arb((0, 1)), RoundUp) = 1.0`. I think this would have more or less well defined semantics and it could actually be quite convenient in some cases. I believe it would make it play nicely with IntervalArithmetic.jl for example. Do you have any thoughts about this? It would sort of be a breaking change, but only because it earlier used the generic Julia implementation for rounded conversion to `Float64`.

It now supports converstion to `Complex{T}` from `Acb` for any `T`, previously it only supported `Float64`. Well, as long as conversion from `Arb` to `T` is supported. The current implementation has one issue/feature, it allows constructing  a `Complex{ArbRef}` which will then not be freshly allocated by refer to the original `Acb` input. In general we don't allow construction of `Ref` types except if one explicitly asks for a reference. I don't think this is a big deal however, most people would not try to construct a `Complex{ArbRef}`.

Regarding handling of complex values we currently have `complex(::Arb)::Complex{Arb}` and `(im * (x::Arb))::Complex{Arb}`. In my current project I deal with a lot of complex numbers and I have realized that it would probably be better to have both of them convert to `Acb` instead of `Complex{Arb}`. But this would be a highly breaking change, so maybe not something we should do at the moment.

### CODE REVIEW

**Comment by Joel-Dahne** [comment]
@OliverHnt  This could possibly be of some interest to you as well. In particular the conversion with rounding to `Float64` and `BigFloat` I realized we were missing when I tried to interface with IntervalArithmetic.jl. The methods `typemin`, `typemax`, `frexp` and `ldexp` I also added because they were used by IntervalArithmetic.jl.

**Comment by Joel-Dahne** [comment]
@kalmarek Would you have time to take a look at this? There is no actually hurry from my side, but would be nice to get it merged!

**Comment by kalmarek** [comment]
@Joel-Dahne I spent the last month traveling and didn't actually open the github page once ;)
I'll look into this tomorrow though, thanks for the patience!

**Comment by Joel-Dahne** [comment]
No worries, take the time you need! I took the opportunity to push to more changes:
1. Fix `Base.precision` for Julia 1.11-DEV. The internal single argument version of Base._precision was renamed to
Base._precision_with_base_2.
2. Add a non-inplace version of `Arblib.load_string!`, for convenience.

I also thought it would make sense to make a release with the updates, so I have bumped the version to 1.1.

**Comment by Joel-Dahne** [comment]
I realized that I had made an error in the conversion tests with `Arblib.arb_rnd`, they should hopefully be fixed now! I accidentally added some additional tests for `BigInt` in the same commit, which I'm now too lazy to move to a new commit.

**Review by kalmarek — CHANGES_REQUESTED** [review]

File: src/conversion.jl [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+end
+Base.Float64(x::ArfOrRef, r::arb_rnd) = get_d(x, r)
+Base.Float64(x::ArbOrRef, r::arb_rnd) = Float64(midref(x), r)
+
+# Deprecated
+Base.Float64(x::ArfOrRef; rnd::arb_rnd = ArbRoundNearest) = Float64(x, rnd)
+
+## Float16 and Float32
+Base.Float16(x::MagOrRef, r::RoundingMode = RoundUp) = Float16(Float64(x, r), r)
+Base.Float16(x::MagOrRef, r::arb_rnd) = Float16(Float64(x, r), convert(RoundingMode, r))
```
[the remark this answers is not in the document]
Comment (Joel-Dahne) [reply src/conversion.jl]: That would certainly be clearer! I updated it to be like that.

File: src/conversion.jl [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    ccall(
+        @libflint(arf_get_fmpz),
+        Cint,
+        (Ref{fmpz_struct}, Ref{arf_struct}, Ref{arb_rnd}),
+        n,
+        x,
+        ArbRoundNearest,
+    )
+    return BigInt(n)
+end
```
[the remark this answers is not in the document]
Comment (Joel-Dahne) [reply src/conversion.jl:89]: I have thought about this as well and it would probably make sense. Will keep this in mind for a future PR!

**Review by kalmarek — APPROVED** [review]

**Pull request merged**